memory/spring-ai-model-chat-memory-jdbc/src/main/java/org/springframework/ai/chat/memory/jdbc/JdbcChatMemoryRepository.java

```diff
@@ -0,0 +1,152 @@
+ * @author Thomas Vitale
+ * @since 1.0.0
+ */
+public class JdbcChatMemoryRepository implements ChatMemoryRepository {
+
+	private static final String QUERY_GET_IDS = """
+			SELECT conversation_id FROM ai_chat_memory
+			""";
+
+	private static final String QUERY_ADD = """
+			INSERT INTO ai_chat_memory (conversation_id, content, type) VALUES (?, ?, ?)
+			""";
+
+	private static final String QUERY_GET = """
+			SELECT content, type FROM ai_chat_memory WHERE conversation_id = ? ORDER BY "timestamp"
+			""";
+
+	private static final String QUERY_CLEAR = "DELETE FROM ai_chat_memory WHERE conversation_id = ?";
+
+	private final JdbcTemplate jdbcTemplate;
+
+	private JdbcChatMemoryRepository(JdbcTemplate jdbcTemplate) {
+		Assert.notNull(jdbcTemplate, "jdbcTemplate cannot be null");
+		this.jdbcTemplate = jdbcTemplate;
+	}
+
+	@Override
+	public List<String> findConversationIds() {
+		List<String> conversationIds = this.jdbcTemplate.query(QUERY_GET_IDS, rs -> {
+			var ids = new ArrayList<String>();
+			while (rs.next()) {
+				ids.add(rs.getString(1));
+			}
+			return ids;
+		});
+		return conversationIds != null ? conversationIds : List.of();
+	}
+
+	@Override
+	public List<Message> findMessages(String conversationId) {
+		Assert.hasText(conversationId, "conversationId cannot be null or empty");
+		return this.jdbcTemplate.query(QUERY_GET, new MessageRowMapper(), conversationId);
+	}
+
+	@Override
+	public void saveMessages(String conversationId, List<Message> messages) {
+		Assert.hasText(conversationId, "conversationId cannot be null or empty");
+		Assert.notNull(messages, "messages cannot be null");
+		Assert.noNullElements(messages, "messages cannot contain null elements");
+		this.jdbcTemplate.batchUpdate(QUERY_ADD, new AddBatchPreparedStatement(conversationId, messages));
+	}
+
+	@Override
+	public void deleteMessages(String conversationId) {
+		Assert.hasText(conversationId, "conversationId cannot be null or empty");
+		this.jdbcTemplate.update(QUERY_CLEAR, conversationId);
+	}
+
+	private record AddBatchPreparedStatement(String conversationId,
+			List<Message> messages) implements BatchPreparedStatementSetter {
+		@Override
+		public void setValues(PreparedStatement ps, int i) throws SQLException {
+			var message = this.messages.get(i);
+
+			ps.setString(1, this.conversationId);
+			ps.setString(2, message.getText());
+			ps.setString(3, message.getMessageType().name());
+		}
+
+		@Override
+		public int getBatchSize() {
+			return this.messages.size();
+		}
+	}
+
+	private static class MessageRowMapper implements RowMapper<Message> {
+
+		@Override
+		@Nullable
+		public Message mapRow(ResultSet rs, int i) throws SQLException {
+			var content = rs.getString(1);
+			var type = MessageType.valueOf(rs.getString(2));
+
+			return switch (type) {
+				case USER -> new UserMessage(content);
+				case ASSISTANT -> new AssistantMessage(content);
+				case SYSTEM -> new SystemMessage(content);
+				case TOOL -> null;
```

Comment:
is returning null ok here?

Reply:
Makes sense! I kept the original logic from JdbcChatMemory, but it does make sense to handle tool messages nicely. I'll do that.

Reply:
I changed it to return an empty ToolResponseMessage, because the storage saves the content of "getText()" and that method returns always an empty string for ToolResponseMessage.

We need a separate enhancement if we want to store the list of tool call results as well.